File: app_manager/appmanager/src/main/java/app/editors/manager/mvp/models/models/ModelExplorerStack.java
package app.editors.manager.mvp.models.models;


import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

import app.editors.manager.app.Api;
import app.editors.manager.mvp.models.explorer.Explorer;
import app.editors.manager.mvp.models.explorer.File;
import app.editors.manager.mvp.models.explorer.Folder;
import app.editors.manager.mvp.models.explorer.Item;

public class ModelExplorerStack {

    private LinkedList<ExplorerStack> mNavigationStack;
    private LinkedList<ExplorerStack> mFilterStack;
    private int mListPosition;

    public ModelExplorerStack() {
        mNavigationStack = new LinkedList<>();
        mFilterStack = new LinkedList<>();
    }

    @Nullable
    public Explorer last() {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            return explorerStack.getExplorer();
        }

        return null;
    }

    @Nullable
    public Explorer previous() {
        final ExplorerStack explorerStack = getPrevious();
        if (explorerStack != null) {
            return explorerStack.getExplorer();
        }

        return null;
    }

    public void addStack(final Explorer explorer) {
        final ExplorerStack explorerStack = new ExplorerStackMap(explorer);

        // First check for filtering
        if (!mFilterStack.isEmpty()) {
            final String currentId = mFilterStack.getLast().getCurrentId();
            if (currentId.equalsIgnoreCase(explorer.getCurrent().getId())) {
                mFilterStack.set(mFilterStack.size() - 1, explorerStack);
            } else {
                mFilterStack.add(explorerStack);
            }
            return;
        }

        // If filtering is empty check navigation stack
        if (!mNavigationStack.isEmpty()) {
            final String currentId = mNavigationStack.getLast().getCurrentId();
            if (currentId.equalsIgnoreCase(explorer.getCurrent().getId())) {
                mNavigationStack.set(mNavigationStack.size() - 1, explorerStack);
                return;
            }
        }

        mNavigationStack.add(explorerStack);
    }

    public void addOnNext(final Explorer explorer) {
        if (!mFilterStack.isEmpty()) {
            mFilterStack.getLast().addExplorer(explorer);
            return;
        }

        if (!mNavigationStack.isEmpty()) {
            mNavigationStack.getLast().addExplorer(explorer);
            return;
        }

        mNavigationStack.add(new ExplorerStackMap(explorer));
    }

    public void refreshStack(final Explorer explorer) {
        checkSelected(explorer);
        if (!mFilterStack.isEmpty()) {
            mFilterStack.getLast().setExplorer(explorer);
            return;
        }

        if (!mNavigationStack.isEmpty()) {
            mNavigationStack.getLast().setExplorer(explorer);
            return;
        }
    }

    public void setFilter(final Explorer explorer) {
        mFilterStack.clear();
        mFilterStack.add(new ExplorerStackMap(explorer));
    }

    public boolean isRoot() {
        if (!mFilterStack.isEmpty()) {
            return false;
        }

        return mNavigationStack.size() < 2;
    }

    public boolean isNavigationRoot() {
        return mNavigationStack.size() < 2;
    }

    public boolean isStackEmpty() {
        return mNavigationStack.isEmpty();
    }

    public boolean isStackFilter() {
        return !mFilterStack.isEmpty();
    }

    public int getTotalCount() {
        int count = 0;
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            count = explorerStack.getCountItems();
        }

        return count;
    }

    public boolean isListEmpty() {
        return getTotalCount() == 0;
    }

    public void setListPosition(final int position) {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            mListPosition = position;
            explorerStack.setListPosition(position);
        }
    }

    public int setSelection(final boolean isSelect) {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            return explorerStack.setSelectionAll(isSelect);
        }
        return 0;
    }

    @Nullable
    public List<String> getSelectedFoldersIds() {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            return explorerStack.getSelectedFoldersIds();
        }
        return null;
    }

    @Nullable
    public List<String> getSelectedFilesIds() {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            return explorerStack.getSelectedFilesIds();
        }
        return null;
    }

    public List<File> getSelectedFiles() {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            return explorerStack.getSelectedFiles();
        }
        return new ArrayList<>();
    }

    public List<Folder> getSelectedFolders() {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            return explorerStack.getSelectedFolders();
        }
        return new ArrayList<>();
    }

    public void addFile(final File file) {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            explorerStack.addFile(file);
        }
    }

    public void addFileFirst(final File file) {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            explorerStack.addFileFirst(file);
        }
    }

    public void addFolder(final Folder folder) {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            explorerStack.addFolder(folder);
        }
    }

    public void addFolderFirst(final Folder folder) {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            explorerStack.addFolderFirst(folder);
        }
    }

    public boolean removeItemById(final String id) {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            return explorerStack.removeItemById(id);
        }

        return false;
    }

    public int removeSelected() {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            return explorerStack.removeSelected();
        }

        return 0;
    }

    public int removeUnselected() {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            return explorerStack.removeUnselected();
        }

        return 0;
    }

    @Nullable
    public Item getItemById(final Item item) {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            return explorerStack.getItemById(item);
        }

        return null;
    }

    @Nullable
    public Item setSelectById(final Item item, final boolean isSelect) {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            return explorerStack.setItemSelectedById(item, isSelect);
        }

        return null;
    }

    public int getCountSelectedItems() {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            return explorerStack.getSelectedItems();
        }

        return 0;
    }

    public int getListPosition() {
        return mListPosition;
    }

    @Nullable
    public String getCurrentTitle() {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            return explorerStack.getCurrentTitle();
        }

        return null;
    }

    @Nullable
    public String getRootId() {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            return explorerStack.getRootId();
        }

        return null;
    }

    public int getRootFolderType() {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            return explorerStack.getRootFolderType();
        }

        return Api.SectionType.UNKNOWN;
    }

    public int getCurrentFolderAccess() {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            return explorerStack.getCurrentFolderAccess();
        }

        return Api.SectionType.UNKNOWN;
    }

    @Nullable
    public String getCurrentFolderOwnerId() {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            return explorerStack.getCurrentFolderOwnerId();
        }

        return null;
    }

    @Nullable
    public String getCurrentId() {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            return explorerStack.getCurrentId();
        }

        return null;
    }

    @Nullable
    public List<String> getPath() {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            return explorerStack.getPath();
        }

        return null;
    }

    public int getLoadPosition() {
        final ExplorerStack explorerStack = getLast();
        if (explorerStack != null) {
            return explorerStack.getLoadPosition();
        }

        return -1;
    }

    @Nullable
    private ExplorerStack getLast() {
        if (!mFilterStack.isEmpty()) {
            final ExplorerStack explorerStack = mFilterStack.getLast();
            mListPosition = explorerStack.getListPosition();
            return explorerStack;
        }

        if (!mNavigationStack.isEmpty()) {
            final ExplorerStack explorerStack = mNavigationStack.getLast();
            mListPosition = explorerStack.getListPosition();
            return explorerStack;
        }

        return null;
    }

    @Nullable
    private ExplorerStack getPrevious() {
        if (!mFilterStack.isEmpty()) {
            mFilterStack.removeLast();
            return getLast();
        }

        if (!mNavigationStack.isEmpty()) {
            mNavigationStack.removeLast();
            if (!mNavigationStack.isEmpty()) {
                return getLast();
            }
        }

        return null;
    }

    @Nullable
    public ExplorerStack clone() {
        if (!mFilterStack.isEmpty()) {
            return mFilterStack.getLast().clone();
        }

        if (!mNavigationStack.isEmpty()) {
            return mNavigationStack.getLast().clone();
        }

        return null;
    }

    public void clear() {
        mFilterStack.clear();
        mNavigationStack.clear();
    }

    private void checkSelected(Explorer explorer) {
        final List<String> selectedFiles = new ArrayList<>();
        final List<String> selectedFolders = new ArrayList<>();

        if (!mFilterStack.isEmpty()) {
            selectedFiles.addAll(mFilterStack.getLast().getSelectedFilesIds());
            selectedFolders.addAll(mFilterStack.getLast().getSelectedFoldersIds());
        } else if (!mNavigationStack.isEmpty()) {
            selectedFiles.addAll(mNavigationStack.getLast().getSelectedFilesIds());
            selectedFolders.addAll(mNavigationStack.getLast().getSelectedFoldersIds());
        }

        if (!mFilterStack.isEmpty() && mFilterStack.getLast().getSelectedItems() > 0) {
            for (File file: explorer.getFiles()) {
                if (selectedFiles.contains(file.getId())) {
                    file.setSelected(true);
                }
            }
            for (Folder folder: explorer.getFolders()) {
                if (selectedFolders.contains(folder.getId())) {
                    folder.setSelected(true);
                }
            }
            return;
        }

        if (!mNavigationStack.isEmpty() && mNavigationStack.getLast().getSelectedItems() > 0) {
            for (File file: explorer.getFiles()) {
                if (selectedFiles.contains(file.getId())) {
                    file.setSelected(true);
                }
            }
            for (Folder folder: explorer.getFolders()) {
                if (selectedFolders.contains(folder.getId())) {
                    folder.setSelected(true);
                }
            }
        }
    }

}

File: app_manager/appmanager/src/main/java/app/editors/manager/mvp/models/models/ModelExplorerStack.kt
package app.editors.manager.mvp.models.models

import app.documents.core.network.common.contracts.ApiContract
import app.documents.core.network.manager.models.explorer.CloudFile
import app.documents.core.network.manager.models.explorer.CloudFolder
import app.documents.core.network.manager.models.explorer.Explorer
import app.documents.core.network.manager.models.explorer.Item
import java.util.LinkedList

class ModelExplorerStack {
    private val navigationStack: LinkedList<ExplorerStack> = LinkedList()
    private val filterStack: LinkedList<ExplorerStack> = LinkedList()
    private var currentListPosition = 0

    fun last(): Explorer? {
        return last?.explorer
    }

    fun previous(): Explorer? {
        return previous?.explorer
    }

    fun popToRoot(): Explorer? {
        return navigationStack.apply {
            val first = navigationStack.first
            clear()
            add(first)
        }.last?.explorer
    }

    fun addStack(explorer: Explorer?) {
        val explorerStack: ExplorerStack = ExplorerStackMap(explorer)

        // First check for filtering
        if (!filterStack.isEmpty() && explorer != null) {
            val currentId = filterStack.last.currentId
            if (currentId.equals(explorer.current.id, ignoreCase = true)) {
                filterStack[filterStack.size - 1] = explorerStack
            } else {
                filterStack.add(explorerStack)
            }
            return
        }

        // If filtering is empty check navigation stack
        if (!navigationStack.isEmpty() && explorer != null) {
            val currentId = navigationStack.last.currentId
            if (currentId.equals(explorer.current.id, ignoreCase = true)) {
                navigationStack[navigationStack.size - 1] = explorerStack
                return
            }
        }
        navigationStack.add(explorerStack)
    }

    fun addOnNext(explorer: Explorer?) {
        if (!filterStack.isEmpty()) {
            filterStack.last.addExplorer(explorer)
            return
        }
        if (!navigationStack.isEmpty()) {
            navigationStack.last.addExplorer(explorer)
            return
        }
        navigationStack.add(ExplorerStackMap(explorer))
    }

    fun refreshStack(explorer: Explorer) {
        checkSelected(explorer)
        if (!filterStack.isEmpty()) {
            filterStack.last.explorer = explorer
            return
        }
        if (!navigationStack.isEmpty()) {
            navigationStack.last.explorer = explorer
            return
        }
    }

    fun setFilter(explorer: Explorer?) {
        filterStack.clear()
        filterStack.add(ExplorerStackMap(explorer))
    }

    val isRoot: Boolean
        get() = if (!filterStack.isEmpty()) {
            false
        } else navigationStack.size < 2

    val isNavigationRoot: Boolean
        get() = navigationStack.size < 2

    val isStackEmpty: Boolean
        get() = navigationStack.isEmpty()

    val isStackFilter: Boolean
        get() = !filterStack.isEmpty()

    val totalCount: Int
        get()  = last?.countItems ?: 0

    val isListEmpty: Boolean
        get() = totalCount == 0

    fun setSelection(isSelect: Boolean): Int {
        return last?.setSelectionAll(isSelect) ?: 0
    }

    val selectedFoldersIds: List<String>
        get() = last?.selectedFoldersIds ?: emptyList()

    val selectedFilesIds: List<String>
        get() = last?.selectedFilesIds ?: emptyList()

    val selectedFiles: List<CloudFile>
        get() = last?.selectedFiles ?: emptyList()

    val selectedFolders: List<CloudFolder>
        get() = last?.selectedFolders ?: emptyList()

    fun addFile(file: CloudFile?) {
        val explorerStack = last
        explorerStack?.addFile(file)
    }

    fun addFileFirst(file: CloudFile?) {
        val explorerStack = last
        explorerStack?.addFileFirst(file)
    }

    fun addFolder(folder: CloudFolder?) {
        val explorerStack = last
        explorerStack?.addFolder(folder)
    }

    fun addFolderFirst(folder: CloudFolder?) {
        val explorerStack = last
        explorerStack?.addFolderFirst(folder)
    }

    fun removeItemById(id: String?): Boolean {
        val explorerStack = last
        return explorerStack?.removeItemById(id) ?: false
    }

    fun removeSelected(): Int {
        val explorerStack = last
        return explorerStack?.removeSelected() ?: 0
    }

    fun removeUnselected(): Int {
        val explorerStack = last
        return explorerStack?.removeUnselected() ?: 0
    }

    fun getItemById(item: Item?): Item? {
        val explorerStack = last
        return explorerStack?.getItemById(item)
    }

    fun setSelectById(item: Item?, isSelect: Boolean): Item? {
        val explorerStack = last
        return explorerStack?.setItemSelectedById(item, isSelect)
    }

    val countSelectedItems: Int
        get() = last?.selectedItems ?: 0

    var listPosition: Int
        get() = currentListPosition
        set(position) {
            val explorerStack = last
            if (explorerStack != null) {
                currentListPosition = position
                explorerStack.listPosition = position
            }
        }

    val currentTitle: String
        get() = last?.currentTitle ?: ""

    val rootId: String?
        get() = last?.rootId

    val rootFolderType: Int
        get() = last?.rootFolderType ?: ApiContract.SectionType.UNKNOWN

    val currentFolderAccess: Int
        get() = last?.currentFolderAccess ?: ApiContract.SectionType.UNKNOWN

    val currentFolderOwnerId: String?
        get() = last?.currentFolderOwnerId

    val currentId: String?
        get() = last?.currentId

    val path: List<String>
        get()  = last?.path ?: emptyList()

    val loadPosition: Int
        get() = last?.loadPosition ?: -1

    private val last: ExplorerStack?
        get() {
            if (!filterStack.isEmpty()) {
                val explorerStack = filterStack.last
                currentListPosition = explorerStack.listPosition
                return explorerStack
            }
            if (!navigationStack.isEmpty()) {
                val explorerStack = navigationStack.last
                currentListPosition = explorerStack.listPosition
                return explorerStack
            }
            return null
        }

    private val previous: ExplorerStack?
        get() {
            if (!filterStack.isEmpty()) {
                filterStack.removeLast()
                return last
            }
            if (!navigationStack.isEmpty()) {
                navigationStack.removeLast()
                if (!navigationStack.isEmpty()) {
                    return last
                }
            }
            return null
        }

    fun clone(): ExplorerStack? {
        if (!filterStack.isEmpty()) {
            return filterStack.last.clone()
        }
        return if (!navigationStack.isEmpty()) {
            navigationStack.last.clone()
        } else null
    }

    fun clear() {
        filterStack.clear()
        navigationStack.clear()
    }

    private fun checkSelected(explorer: Explorer) {
        val selectedFiles: MutableList<String> = ArrayList()
        val selectedFolders: MutableList<String> = ArrayList()
        if (!filterStack.isEmpty()) {
            selectedFiles.addAll(filterStack.last.selectedFilesIds)
            selectedFolders.addAll(filterStack.last.selectedFoldersIds)
        } else if (!navigationStack.isEmpty()) {
            selectedFiles.addAll(navigationStack.last.selectedFilesIds)
            selectedFolders.addAll(navigationStack.last.selectedFoldersIds)
        }
        if (!filterStack.isEmpty() && filterStack.last.selectedItems > 0) {
            for (file in explorer.files) {
                if (selectedFiles.contains(file.id)) {
                    file.isSelected = true
                }
            }
            for (folder in explorer.folders) {
                if (selectedFolders.contains(folder.id)) {
                    folder.isSelected = true
                }
            }
            return
        }
        if (!navigationStack.isEmpty() && navigationStack.last.selectedItems > 0) {
            for (file in explorer.files) {
                if (selectedFiles.contains(file.id)) {
                    file.isSelected = true
                }
            }
            for (folder in explorer.folders) {
                if (selectedFolders.contains(folder.id)) {
                    folder.isSelected = true
                }
            }
        }
    }

}